init.lua

```diff
@@ -842,6 +842,51 @@ function Form:set_as_inventory_for(player, ctx)
     player:set_inventory_formspec(fs)
 end
 
+-- Declared here to be accessible by render_to_formspec_string
+local fs_process_events
+
+-- Prevent collisions in forms, but also ensure they don't happen across
+-- mutliple embedded forms within a single parent.
+-- Unique per-user to prevent players from making the counter wrap around for
+-- other players.
+local render_to_formspec_auto_name_ids = {}
+-- if `standalone` is not true, then this won't output a "root" formspec, but
+--   instead will output  formspec designed to be embedded along with the
+--   information that was removed from the form
+function Form:render_to_formspec_string(player, ctx, standalone)
+    local name = player:get_player_name()
+    local info = minetest.get_player_information(name)
+    local tree, form_info = self:_render(player, ctx or {},
+        info and info.formspec_version, render_to_formspec_auto_name_ids[name])
+    local public_form_info
+    if standalone then
+        local size = table.remove(tree, 1)
+        public_form_info = {w = size.w, h = size.h,
+            formspec_version = tree.formspec_version}
+        tree.formspec_version = nil
+    end
+    local fs = assert(formspec_ast.unparse(tree))
+    render_to_formspec_auto_name_ids[name] = form_info.auto_name_id
+    local function event(fields)
+        -- Just in case the player goes offline, we should not keep the player
+        -- reference. Nothing prevents the user from calling this function when
+        -- the player is offline, unlike the _real_ formspec submission.
+        local player = minetest.get_player_by_name(name)
+        if not player then
+            minetest.log("warning", "[flow] Player " .. name ..
+                " was offline when render_to_formspec_string event was" ..
+                " triggered. Events were not passed through.")
+            return nil
+        end
+        return fs_process_events(player, form_info, fields)
+    end
+    if standalone then
+        return fs, event, public_form_info
+    else
+        return fs, event
+    end
```

Comment:
The if/else statement could be removed entirely, though then it'd return `fs, event, nil` when standalone is set.

```suggestion
    return fs, event, public_form_info
```

Reply:
Added in latest commit.